File: linked list/Merge k Sorted Lists.py
from pydantic import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def merge(left: Optional[ListNode], right: Optional[ListNode]) -> Optional[ListNode]:
    merged = ListNode()
    curr = merged
    while left and right:
        if left.val < right.val:
            curr.next = left
            left = left.next
        else:
            curr.next = right
            right = right.next
        curr = curr.next

    if left:
        curr.next = left
    if right:
        curr.next = right

    return merged.next


def mergeKLists(lists: list[Optional[ListNode]]) -> Optional[ListNode]:
    if not lists:
        return None
    if len(lists) == 1:
        return lists[0]

    mid = len(lists) // 2
    left = mergeKLists(lists[:mid])
    right = mergeKLists(lists[mid:])

    return merge(left, right)

File: linked list/Merge k Sorted Lists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class SolutionMergeKLists
{
    private ListNode Merge(ListNode m1, ListNode m2)
    {
        ListNode merged = new ListNode(0);
        ListNode curr = merged;
        while (m1 != null && m2 != null)
        {
            if (m1.val <= m2.val)
            {
                curr.next = m1;
                m1 = m1.next;
            }
            else
            {
                curr.next = m2;
                m2 = m2.next;
            }
            curr = curr.next;
        }

        if (m1 != null)
            curr.next = m1;
        else
            curr.next = m2;

        return merged.next;
    }
    public ListNode MergeKLists(ListNode[] lists)
    {
        int k = lists.Length;
        if (k == 0)
            return null;

        for (int i = 1; i < lists.Length; ++i)
        {
            ListNode merged = Merge(lists[0], lists[i]);
            lists[0] = merged;
        }

        return lists[0];
    }
}
